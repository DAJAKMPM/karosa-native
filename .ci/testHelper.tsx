import React, { JSXElementConstructor } from "react";
import { View } from "react-native";
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from "@testing-library/react-native";
import { Formik, FormikValues } from "formik";

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  _themeProp: unknown = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };

  const renderApi = render(children, options);
  const wrapper = renderApi.queryByTestId(wrapperTestID || "wrapper");
  return { wrapper, ...renderApi };
};

export const renderWithFormikWrapper = <T extends FormikValues>(
  initialValues: T,
  children: React.ReactElement<any, string | JSXElementConstructor<any>>
) => {
  return renderWithWrapper(
    <Formik
      initialValues={initialValues}
      onSubmit={() => console.log("submit")}
      component={() => children}
    />
  );
};
