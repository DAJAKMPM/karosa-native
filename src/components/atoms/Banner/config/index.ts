/**
 *
 * @format
 *
 */
export const INACTIVEDOTOPACITY = 0.4
export const INACTIVESCALEDOT = 0.6
export const LAYOUTCARDOFFSET = 0



