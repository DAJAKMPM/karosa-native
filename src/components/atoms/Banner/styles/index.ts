/**
 *
 * @format
 *
 */

import { DIMENS } from "@app/styles";
import { StyleSheet } from "react-native";

export const SLIDER_WIDTH = DIMENS.screenWidth;
export const ITEM_WIDTH = Math.round(SLIDER_WIDTH * 0.7);

const BannerStyles = StyleSheet.create({
  image: {
    width: "100%",
    height: 300,
  },
  carouselItem: {
    backgroundColor: "white",
    borderRadius: 8,
    width: ITEM_WIDTH,
    paddingBottom: 40,
    shadowColor: "#000",
    shadowOffset: {
      width: 0,
      height: 3,
    },
    shadowOpacity: 0.29,
    shadowRadius: 4.65,
    elevation: 7,
  },
  carouselContainer: {
    width: "100%",
  },
  paginationStyle: {
    position: "absolute",
    flexDirection: "column-reverse",
    height: "100%",
    alignSelf: "center",
  },
  dotStyle: {
    width: 10,
    height: 10,
    borderRadius: 5,
    marginHorizontal: 0,
    backgroundColor: "white",
  },
});

export default BannerStyles;
