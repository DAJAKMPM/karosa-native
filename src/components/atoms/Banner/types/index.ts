/**
 *
 * @format
 *
 */

export type CarouselData = {
  source: string;
};
export type PropsType = {
  carouselData: CarouselData[];
};
