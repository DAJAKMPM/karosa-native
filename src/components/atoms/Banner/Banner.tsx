/**
 *
 * Banner
 * @format
 *
 */

import React, { FC, useState, useRef } from "react";
import { Image, View } from "react-native";
import { DIMENS } from "@app/styles";
import Carousel, { Pagination } from "react-native-snap-carousel";
import {
  INACTIVEDOTOPACITY,
  INACTIVESCALEDOT,
  LAYOUTCARDOFFSET,
} from "./config";

import type { PropsType, CarouselData } from "./types";
import BannerStyles from "./styles";

const Banner: FC<PropsType> = (props) => {
  const { carouselData } = props;
  const SLIDER_WIDTH = DIMENS.screenWidth * 0.9;

  const renderImage = (item: { item: CarouselData }) => {
    return (
      <View>
        <Image source={{ uri: item.item.source }} style={BannerStyles.image} />
      </View>
    );
  };

  const [index, setIndex] = useState(0);
  const carouselRef = useRef<Carousel<CarouselData>>(null);

  return (
    <View style={BannerStyles.bannerContainer}>
      <Carousel
        layout="stack"
        data={carouselData}
        ref={carouselRef}
        renderItem={renderImage}
        sliderWidth={SLIDER_WIDTH}
        itemWidth={SLIDER_WIDTH}
        layoutCardOffset={LAYOUTCARDOFFSET}
        onSnapToItem={(value) => setIndex(value)}
      />
      <View style={BannerStyles.paginationStyle}>
        <Pagination
          dotsLength={carouselData.length}
          activeDotIndex={index}
          dotStyle={BannerStyles.dotStyle}
          inactiveDotOpacity={INACTIVEDOTOPACITY}
          inactiveDotScale={INACTIVESCALEDOT}
        />
      </View>
    </View>
  );
};

export default Banner;
