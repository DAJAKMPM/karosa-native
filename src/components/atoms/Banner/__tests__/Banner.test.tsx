import React from "react";
import { renderWithWrapper } from "@app/ci/testHelper";
import Banner from "../Banner";

const carouselList = [
  {
    source: "https://picsum.photos/200/300",
  },
  {
    source: "https://picsum.photos/200/300",
  },
  {
    source: "https://picsum.photos/200/300",
  },
];

global.requestAnimationFrame = function (callback) {
  return setTimeout(callback, 0);
};

global.cancelAnimationFrame = function (id) {
  clearTimeout(id);
};

describe("Banner Component", () => {
  it("should match snapshot", () => {
    const { toJSON } = renderWithWrapper(
      <Banner carouselData={carouselList} />
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
