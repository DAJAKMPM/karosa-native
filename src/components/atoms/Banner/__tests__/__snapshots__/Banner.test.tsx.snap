// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner Component should match snapshot 1`] = `
<View
  testID="wrapper"
>
  <View
    style={
      Object {
        "overflow": "hidden",
        "position": "relative",
      }
    }
  >
    <RCTScrollView
      activeAnimationOptions={null}
      activeAnimationType="timing"
      activeSlideAlignment="center"
      activeSlideOffset={20}
      apparitionDelay={0}
      automaticallyAdjustContentInsets={false}
      autoplay={false}
      autoplayDelay={1000}
      autoplayInterval={3000}
      callbackOffsetMargin={5}
      containerCustomStyle={Object {}}
      contentContainerCustomStyle={Object {}}
      contentContainerStyle={
        Array [
          Object {
            "paddingLeft": 0,
            "paddingRight": 0,
          },
          Object {},
        ]
      }
      data={
        Array [
          Object {
            "source": "https://picsum.photos/200/300",
          },
          Object {
            "source": "https://picsum.photos/200/300",
          },
          Object {
            "source": "https://picsum.photos/200/300",
          },
        ]
      }
      decelerationRate="fast"
      directionalLockEnabled={true}
      enableMomentum={false}
      enableSnap={true}
      firstItem={0}
      forwardedRef={[Function]}
      hasParallaxImages={false}
      horizontal={true}
      inactiveSlideOpacity={0.7}
      inactiveSlideScale={0.9}
      inactiveSlideShift={0}
      inverted={false}
      itemWidth={675}
      layout="stack"
      layoutCardOffset={0}
      lockScrollTimeoutDuration={1000}
      lockScrollWhileSnapping={false}
      loop={false}
      loopClonesPerSide={3}
      onLayout={[Function]}
      onResponderRelease={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      onSnapToItem={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      overScrollMode="never"
      pinchGestureEnabled={false}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEnabled={true}
      scrollEventThrottle={1}
      scrollsToTop={false}
      shouldOptimizeUpdates={true}
      showsHorizontalScrollIndicator={false}
      showsVerticalScrollIndicator={false}
      slideStyle={Object {}}
      sliderWidth={675}
      style={
        Object {
          "flexDirection": "row",
          "opacity": 0,
          "width": 675,
        }
      }
      swipeThreshold={20}
      useScrollView={false}
      vertical={false}
    >
      <View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "opacity": 1,
              "transform": Array [
                Object {
                  "scale": 1,
                },
                Object {
                  "translateX": 0,
                },
              ],
              "width": 675,
              "zIndex": 3,
            }
          }
        >
          <View>
            <Image
              resizeMode="cover"
              source={
                Object {
                  "uri": "https://picsum.photos/200/300",
                }
              }
              style={
                Object {
                  "borderRadius": 10,
                  "height": 180,
                  "width": "100%",
                }
              }
            />
          </View>
        </View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "opacity": 0.75,
              "transform": Array [
                Object {
                  "scale": 0.9,
                },
                Object {
                  "translateX": -716,
                },
              ],
              "width": 675,
              "zIndex": 2,
            }
          }
        >
          <View>
            <Image
              resizeMode="cover"
              source={
                Object {
                  "uri": "https://picsum.photos/200/300",
                }
              }
              style={
                Object {
                  "borderRadius": 10,
                  "height": 180,
                  "width": "100%",
                }
              }
            />
          </View>
        </View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "opacity": 0.5,
              "transform": Array [
                Object {
                  "scale": 0.8,
                },
                Object {
                  "translateX": -1620,
                },
              ],
              "width": 675,
              "zIndex": 1,
            }
          }
        >
          <View>
            <Image
              resizeMode="cover"
              source={
                Object {
                  "uri": "https://picsum.photos/200/300",
                }
              }
              style={
                Object {
                  "borderRadius": 10,
                  "height": 180,
                  "width": "100%",
                }
              }
            />
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "alignSelf": "center",
          "flexDirection": "column-reverse",
          "height": "100%",
          "position": "absolute",
        }
      }
    >
      <View
        accessible={false}
        pointerEvents="box-none"
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "paddingHorizontal": 20,
              "paddingVertical": 30,
            },
            Object {
              "flexDirection": "row",
            },
            Object {},
          ]
        }
      >
        <View
          accessible={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "marginHorizontal": 8,
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#fff",
                "borderRadius": 5,
                "height": 10,
                "marginHorizontal": 0,
                "opacity": 0.4,
                "transform": Array [
                  Object {
                    "scale": 0.6,
                  },
                ],
                "width": 10,
              }
            }
          />
        </View>
        <View
          accessible={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "marginHorizontal": 8,
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#fff",
                "borderRadius": 5,
                "height": 10,
                "marginHorizontal": 0,
                "opacity": 0.4,
                "transform": Array [
                  Object {
                    "scale": 0.6,
                  },
                ],
                "width": 10,
              }
            }
          />
        </View>
        <View
          accessible={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "marginHorizontal": 8,
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#fff",
                "borderRadius": 5,
                "height": 10,
                "marginHorizontal": 0,
                "opacity": 0.4,
                "transform": Array [
                  Object {
                    "scale": 0.6,
                  },
                ],
                "width": 10,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
